chain/default_syncer.go

```diff
@@ -371,6 +371,7 @@ func (syncer *DefaultSyncer) HandleNewBlocks(ctx context.Context, blkCids []cid.
 			}
 		}
 		if err = syncer.syncOne(ctx, parent, ts); err != nil {
+			syncer.badTipSets.AddChain(chain[i:])
```

Comment:
blocking: this doesn't differentiate between a bad tipset error and an error because something didn't work as intended. seems like it should be an important distinction, is it not?

Reply:
@rkowalick and I talked a lot about this and we landed on this not being an important distinction, at least not right now.  

There is exactly one expected return error from `syncOne` during normal operation: https://github.com/filecoin-project/go-filecoin/blob/master/chain/default_syncer.go#L219.  If this error is returned it means this tipset does not conform to the consensus rules (invalid leader election proof, invalid state machine transition, etc).  In this case the tipset will always be bad and the chain based on this tipset should be marked as bad.

In addition to this one error there are many other errors that can be returned in the case that fundamental things are broken.  For example most of these will be thrown if the chain store is in a corrupt state.  These errors are like the "fault" errors explicitly denoted in the state machine processing code.  

Our reasoning for marking the tipset bad in this case is that we think the behavior of the node is undefined in the case there is a chain store fault, so it's not important to *not* add it to the cache in this case.  Additionally a situation like that is going to require a node restart (if not re-init) and that will clear the bad tipsets cache anyway.

This code can be modified to check the error and only add to the cache in that case.  I'm thinking the syncer code should also be more explicit about chain store faults and try to do something about them.  Probably a good start is to log them very loudly and explicitly.

Reply:
Please summarise this comment in the code.

Reply:
Thanks @ZenGround0. I buy the logic that's it's not a bad thing to mark a tip set as bad under fault conditions. 

What's still unclear to me as a relative newcomer to this code is why we're doing this in the first place. It appears the badTipSets cache prevents us from having to re-validate tip sets we've determined bad (either by being bad or by having a bad parent) in `collectChain`. It appears that we are expanding the definition of "bad" to include tip sets that fail to sync. Questions (cc @rkowalick):

1. How does this help resolve #2155? The issue seems to have been closed without this change.
2. Is this more than a performance optimization?
3. If a user has a long running node and encounters a hiccup syncing a valid block that creates a false positive here, how do they know to restart the node?

Reply:
1. This behavior of seeing a chain syncing over and over in the logs came up as complaint which launched investigation of #2155.  @rkowalick figured out that the root cause was consensus state transition error and made this change to improve the UX of the failure.  I agree this should refer to a follow on and not #2155 itself.
2. No it does not change correctness.  It does makes logs less confusing.
3. Right now they wouldn't.  A power user might infer that behavior from seeing the `the chain store is in an unexpected state` in the logs.

Reply:
I have added a comment in the `badTipSetCache` to clarify the purpose of the cache for the future. I am apprehensive about adding a comment at this call site, since, according to @ZenGround0, the change above certainly agrees with the original motivation behind this cache.